doctr/models/recognition/sar/pytorch.py

```diff
@@ -93,65 +99,69 @@ def __init__(
         embedding_units: int,
         attention_units: int,
-        num_decoder_layers: int = 2,
         feat_chans: int = 512,
         dropout_prob: float = 0.,
     ) -> None:
 
         super().__init__()
         self.vocab_size = vocab_size
-        self.rnn = nn.LSTM(rnn_units, rnn_units, 2, batch_first=True, dropout=dropout_prob)
-        self.embed = nn.Embedding(self.vocab_size + 1, embedding_units)
-        self.attention_module = AttentionModule(feat_chans, rnn_units, attention_units)
-        self.output_dense = nn.Linear(2 * rnn_units, vocab_size + 1)
         self.max_length = max_length
 
+        self.embed = nn.Linear(self.vocab_size + 1, embedding_units)
+        self.embed_tgt = nn.Embedding(embedding_units, self.vocab_size + 1)
+        self.attention_module = AttentionModule(feat_chans, rnn_units, attention_units)
+        self.lstm_cell = nn.LSTMCell(rnn_units, rnn_units)
+        self.output_dense = nn.Linear(2 * rnn_units, self.vocab_size + 1)
+        self.dropout = nn.Dropout(dropout_prob)
+
     def forward(
         self,
         features: torch.Tensor,  # (N, C, H, W)
         holistic: torch.Tensor,  # (N, C)
-        gt: Optional[torch.Tensor] = None,  # (N, L)
+        gt: Optional[torch.Tensor] = None  # (N, L)
     ) -> torch.Tensor:
 
-        if gt is None:
-            # Initialize with the index of virtual START symbol (placed after <eos> so that the one-hot is only zeros)
-            symbol = torch.zeros(features.shape[0], device=features.device, dtype=torch.long)
-            # (N, embedding_units)
-            symbol = self.embed(symbol)
-            # (N, L, embedding_units)
-            symbol = symbol.unsqueeze(1).expand(-1, self.max_length + 1, -1)
-        else:
-            # (N, L) --> (N, L, embedding_units)
-            symbol = self.embed(gt)
-        # (N, L + 1, embedding_units)
-        symbol = torch.cat((holistic.unsqueeze(1), symbol), dim=1)
-
-        # (N, L, vocab_size + 1)
-        char_logits = torch.zeros(
-            (features.shape[0], self.max_length + 1, self.vocab_size + 1),
-            device=features.device,
-            dtype=features.dtype,
-        )
-
-        decoding_iter = self.max_length + 1 if gt is None else 1
-        for t in range(decoding_iter):
-            # (N, L + 1, rnn_units)
-            logits = self.rnn(symbol)[0]
-            # (N, L + 1, C)
-            glimpse = self.attention_module(features, logits)
-            # (N, L + 1, 2 * rnn_units)
-            logits = torch.cat((logits, glimpse), -1)
-            # (N, L + 1, vocab_size + 1)
-            decoded = self.output_dense(logits)
-            if gt is None:
-                char_logits[:, t] = decoded[:, t + 1]
-                if t < decoding_iter - 1:
-                    # update symbol with predicted logits for t + 1 step
-                    symbol[:, t + 2] = self.embed(decoded[:, t + 1].argmax(-1))
+        if gt is not None:
+            gt_embedding = self.embed_tgt(gt)
```

Comment:
I think I see that part of the previous mistake was here, thanks for fixing it!

Reply:
Yes before it was (N, embedding_units) but i think there was also a problem with indexing :)